### ISSUE TITLE
Adding diacritics to unpointed Arabic text 

### ISSUE THREAD

**Issue by AhMohsen46** [interscript/interscript-ruby#719]
Diacritics in Arabic are Latin short vowels.
examples: 
َ  : a
ُ  : w/o
ِ  : i/e
ّ  : doubling the letter
ْ  : omitted 
Mostly, in Arabic text, diacritics are omitted from writing, but still affecting the pronunciation, and consequently, the transliteration of it to Latin script
example:
فلسطين‎ is transliterated to Filasṭīn
ف = F
ل = L
س = S
ط = T
ي = Y
ن = N

note that if this is transliterated, it'll be FLSTIN, while the correct transliteration is Filasṭīn

the correct arabic text with pointing should have been فِلَسْطِين 
ِ  should be converted to i after the F
َ  should be converted to a after the L
ْ  is omitted
and finally,  ِ  before the ي makes it a long vowel, so instead of i it's transliterated to ii, which is the converted to ī

although currently have most of the available-to-implement Arabic maps implemented, we still have a challenge left, which is converting non-pointed text to pointed text (diacritizing it)

another issue in pointing is collisions between words, with the same characters if diacritics are omitted
example: 
قَطَر (Qatar - a country)
قُطْر (Qotr - circle diameter)
قَطْر (Qatr - rain)
and others, where the context might be more helpful than the word on its own

an approach to do that is dealing with un-pointed and pointed arabic as two different languages, and convert the unpointed to pointed using seq2seq model in [Secryst](https://github.com/secryst/secryst)

we prepared a dataset of almost 250k unique words [here](https://github.com/secryst/data-arabic-pointing), we have data with collisions, and without collisions

but unfortunately, the model failed in 2 things:

1-the output was not the same length as the input after omitting the diacritics -- how I believe this should be fixed, is by making the criterion function remove the diacritics, and validate that the lengths stay the same after pointing
so for the word قطر it's not acceptable to have it transformed to قَطَرا or anything like that, since after omitting  َ قطر length != قطرا length

2-handling collisions, ideally, I think it might be decided based on the context, to start with, we may try to detect names, which we're more concerned with for the initial phase, @ronaldtse your opinion here would be appreciated, thank you

----

some resources for similar ML models and some papers (links included within the repos) focused on arabic diacritization can be found here:

https://github.com/Barqawiz/Shakkala

as a dataset, this is the one i think he used in the publication: https://github.com/AliOsm/arabic-text-diacritization

also, I found this, https://alt.qcri.org/farasa, an open api for non-commercial use

and, this python package: https://github.com/linuxscout/mishkal is great, but am not sure if it'll force installing the neural network

that's how it works: 

Mishkal use a rule based method to detect relations and diacritics, First, it analyzes all morphological cases, it generates all possible diacritized word forms, by detecting all affixes and check it in a dictionary. second, It add word frequency to each word.

The two previous steps are made by support/Qalsadi ( arabic morphological analyzer), the used dictionary is a separated project named 'Arramooz: arabic dictionnary for morphology".

Third, we use a syntax analyzer to detect all possible relations between words. The syntax library is named support/ArAnaSyn. This analyzer is basic for the moment, it use only linear relations between adjacent words.

Forth, all data generated and relations will be analyzed semantically, to detect semantic relation in order to reduce ambiguity. The use libary is support/asmai ( Arabic semantic analysis). The semantic relations extraction is based on corpus. The used corpus is named "Tashkeela: arabic vocalized texts corpus".

In the final stage, The module mishkal/tashkeel tries to select the suitable word in the context, it tries to get evidents cases, or more related cases, else, it tries to select more probable case, using some rules like select a stop word by default, or select Mansoub case by default.

----

we've a few other datasets that we found while searching, one of them has more than 70m pointed words (non-unique), that can be converted to a dataset by unpointing them using a regex or something, will get the links to those soon

**Comment by ronaldtse** [comment]
Regarding @AhMohsen46 's comment about fixed-length outputs, my potential solution was here:
https://github.com/secryst/data-arabic-pointing/issues/1

**Comment by ronaldtse** [comment]
Ping @gilgameshjw this is the task at hand.

**Comment by gilgameshjw** [comment]
### Summary
Hello, over the last days, I have reviewed some literature, mainly this [article](https://arxiv.org/pdf/1905.01965.pdf). I have also run tests with **shakkala** and **mishkal** libraries on both yours and the shakkala benchmark datasets. 

### Findings on diacritisation
1. [shakkala](https://github.com/Barqawiz/Shakkala) library not only crushes other libraries, but it is relatively easy to run. I have also checked that I could replicate the benchmarks of table V in [article](https://arxiv.org/pdf/1905.01965.pdf) within few hours. They do have a [live demo](https://ahmadai.com/shakkala/). 
3. **metrics and benchmarking:** as mentioned by **[AhMohsen46](https://github.com/AhMohsen46)**, datasets and code for benchmarks can be found on [diacritisation-metrics](https://github.com/AliOsm/arabic-text-diacritization). **NER** and **WER** are easy to understand as errors rates on diacritics and words respectively. 
4. **Longer outputs:** If I have understood things correctly, they have fixed that issue. It is worthwhile to mention Fig.4 of [article](https://arxiv.org/pdf/1905.01965.pdf), mentioning that precise issue as **ending diacritics**, which they claim they fix during the preprocessing, prior to training with code available in [diacritisation-metrics](https://github.com/AliOsm/arabic-text-diacritization). The function is **fix_diacritics** defined in the notebook, in [helpers](https://github.com/AliOsm/arabic-text-diacritization/blob/master/helpers/pre_process_tashkeela_corpus.ipynb).  
5. **tests on your datasets:** I have run both **mishkal** and **shakkala** on a fraction of [your dataset](https://github.com/secryst/data-arabic-pointing).  For both vocabularies, collisions and without, shakkala and mishkal success rates at guessing all diacritics correctly (but without any context!) are resp. $\approx$ 30%/22%(shakkala) and 0%/0%(!?mishkal). Both scores came at a surprise given that shakkala features good scores. It points out at some misunderstanding, probably on my side... of what with collisions and without mean. But I would have expected much higher scores, so it is important to check!

### Suggestions moving forward
**Shakkala** features great scores compared to previous softwares.  Whether it is good enough for your business cases is what you need to check.  If this were the case, you could just overtake their library...

If not, then you need to benchmark your library and ideas with the metrics and datasets mentioned above.  
This would become an involved project, but shakkala architecture and dataset could be improved. They are probably the options of doing it yourself or even with the Shakkala team because these guys are following a consistent open source policy. 

They aren't so many teams that have tried to improve diacritization, it seems to me (this should be checked with a bit more research within open source and non open source ecosystems). So maybe you want to evaluate the costs/opportunity of doing that...

Preprocessing your algo. outputs or the training data might just sort out the problem of wrong lengths. Ronald mentioned an idea for the seq2seq training that sounds reasonable but is one of the numerous ideas one would need to try and test against the solid benchmarks mentioned above.

**Comment by AhMohsen46** [comment]
@gilgameshjw Thanks for the effort investigating!
1-I'm in favor of shakkala, it's also the most widely used of the available ones, most Started on Github too, but I remember a discussion I had with @ronaldtse  about it, and he said that we wanna avoid installing neural network libraries and have it in ONNX format instead, and I frankly didn't have enough knowledge in Neural networks to talk about that back then

3-For longer outputs, so, what I initially meant, was using secryst, which we already have, that's definitely a great find and can be helpful for us in fixing it in secryst, if we'll end up using it and won't use any alternative, thank you!

4-a)I'm sorry I don't understand the scores: $\approx$ 30%/22%(shakkala) and 0%/0%(!?mishkal)., do these numbers stand for success or failures?, can you explain further please, thank you

4-b)About collisions, will try to re-phrase the explanation
In Arabic, diacritics, can change the wordd meaning, and they really differ
here's an example: 
![image](https://user-images.githubusercontent.com/34799755/120940653-a53b3800-c71e-11eb-91cc-5242d69bc322.png)
here, you can see a word like عقد, has 6 different meaning in English, depending on the diacritics
now that our system purpose is pointing arabic text
for the word as example, how can we decide whether it's عَقْد (contract) or عُقْد (necklace)?
pretty complicated I know
even me myself, as a native Arabic speaker, if I saw this word without context, I'll face difficulty pronouncing it
unless I see it in context like: اتشريت عقد, this means, `I bought a necklace`, so I'll know it's عُقد for certain
while if it was like: وقّعت عقد, this means, `I've signed a contract`, I'll know it's عَقْد
one last example: الرجل عقد الأمور, this means, `the man has complicated things`, I'll know it's عَقَّدَ


for example, in our case, since we're using it for nouns, we'll first eliminate عَقَّدَ from our choices, then if there's a context, we'll decide based on it as explained above

----

another example that we've just started trying yesterday is this:
 https://github.com/linuxscout/mishkal

JSON.parse(HTTParty.get("http://tahadz.com/cgi-bin/mishkal.cgi/ajaxGet?text=%D9%81%D9%84%D8%B3%D8%B7%D9%8A%D9%86&action=Tashkeel2&order=0&lastmark=1").body)
{
    "result" => [
        [0] {
               "semi" => "فِلَسْطِين",
             "chosen" => "فِلَسْطِينٌ",
            "suggest" => "فَلَسُطِينَ;فِلَسْطِينٌ;فِلَسْطِينٍ;فِلَسْطِينَ;فِلَسْطِينُ;فِلَسْطِينِ",
            "inflect" => "Noun:جامد:::مرفوع:متحرك:تنوين:::T4G0N1",
               "rule" => 31,
               "link" => ""
        }
    ],
     "order" => "0"
}

the good part, is it's rule based, so I think it might be relatively faster

we can either use the endpoint, and have our own cache for frequently used words to hit before trying to fetch the data from the endpoint

or, we may just use the whole package, if it sounds like a valid option

**Comment by gilgameshjw** [comment]
A few things to mention here:
* I could run shakkala on my slow cpu's...
*  it was a few times faster than mishkal to run on realistic examples of the dataset. 
* Also, you don't need to train anything, the model can be loaded.
* Accuracy is much better  (see article above, 6% vs 26% WER... for Shakkala vs Mishkal...)

Their is a complication with large sentences that would need to be cut into smaller pieces but all the code for doing that is available. 

If you prefer mishkal, you could just cut the endings with some of the code that I have highlighted, but it is a hack.

Thx for comments on collisions. scores are just the number of correct diacrisations on words from your data set. shakkala struggles (30% correct on non collision dataset??) but mishkal fails completely (0%) on the words without collisions. 
I could append samples of predictions if useful?

**Comment by gilgameshjw** [comment]
* In term of WER it means 70% for shakkala and 100% for mishkal on the dataset without collision...?
* If the pbm is the deployment of NNs, there are container technologies. My guess is that anything not NNs is not near to do the job compared to NNs here. I could help with that.
* I don't know the business case though... What you need depend of that...

**Comment by ronaldtse** [comment]
Our goal is to have the "diacritiziation model" realized in ONNX, which allows the user to directly utilize the model without any development ML/NN code installed (e.g. libtorch).

It looks like Shakkala (at least its dataset) is the way to go due to the low error rate as they demonstrated (also better than Mishkal):
<img width="778" alt="Screenshot 2021-06-07 at 12 32 12 PM" src="https://user-images.githubusercontent.com/11865/120959465-65ab3500-c78c-11eb-93f0-d5ef0292635e.png">

@gilgameshjw can you help:

1. Port Shakkala from Tensorflow to PyTorch so we can export it into an ONNX model
2. Potentially improve on the error rates

?

In addition, we also need a way to find out (during or outside of diacriticization) to know which phrases are proper nouns (place names / personal names). For example, some place names are named after a date. @AhMohsen46 we may need to find some dictionary to do so?

Thanks.

**Comment by gilgameshjw** [comment]
Hello,

1. **model to pytorch:** perfectly doable given that the model precise specs are available if you load the model. I think it is around one week at the current rate to get the model replicated (complications are expected with arabic and preprocessing), maybe not with the same rates (I have only access to a 8 cores CPU machine currently, which might be enough but slow...).
2.  **improvements:** we can try for another week or two after 1. I would need access to GPU's to make experiments. We have a few ideas and potentially additional data?
3. proper nouns (place names / personal names): 
    * personal names: is there a data set with that or some huge file with names that we could scrape? 
    * location names: www.openstreetmap.org it should be possible to query it with Long Lat and farm the location names that way. Maybe there are better ideas for Arabia? How granular is the naming on openstreetmap?
   
@ronaldtse @AhMohsen46 Is the schedule sketched above ok?
Can you give me access to some GPU's for the second--third weeks?

**Comment by ronaldtse** [comment]
> 1. model to pytorch

That's excellent. Let's try out with what 

> 2. improvements

GPU: of course, we can provide AWS access.

> 3. proper nouns
> * is there a data set with that or some huge file with names that we could scrape

Probably @AhMohsen46 would have an idea.

> * location names: www.openstreetmap.org it should be possible to query it with Long Lat and farm the location names that way. Maybe there are better ideas for Arabia? How granular is the naming on openstreetmap?

The database to use here is the [GeoNames database](http://geonames.nga.mil), it is accurate as I understand.

**Comment by gilgameshjw** [comment]
@AhMohsen46  @ronaldtse I found and read new ressources, **2020 benchmarks!**:
[article](https://arxiv.org/pdf/1905.01965.pdf) and [code](https://github.com/Hamza5/Pipeline-diacritizer)

**Comment by ronaldtse** [comment]
This looks really impressive! The code isn’t long either. 

This particular system claims to work better than Shakkala, and uses Tensorflow together with some post-DL cleaning. We could implement the training part with PyTorch, and maybe @gilgameshjw can improve on the results as well, which I suspect can be done. (In this paper they separate the shadda as a second dimension, but maybe it doesn’t need to be — @AhMohsen46 does adding of the shadda have anything to do with the other diacritics?). 

The results show immense improvement in post induction rule-based cleaning (removing diacritics that have obvious errors or correcting to known words). I wonder if there is a cleaner way of doing so.

**Comment by gilgameshjw** [comment]
We just had a long chat with @AhMohsen46 .

As we discover, it seems that this code (unlike shakkala) is complete and has the data encoding and preprocessing bits made available. It looks like one can just take the tensorflow NNets and just transform it into its pytorch equivalent.

If that were to go without problem as it should, we discussed about the most straightforward improvements we could make:
Here we discussed about transformers [arabic BERT](https://github.com/aub-mind/arabert) and attention VS LSTM. We think that this would have reasonable chances (backed up by research results in NLP) to improve results on an existing, advanced solution.
Alongside of this, we can try other encoding as @ronaldtse  suggests.

Special cases to handle are when sentences are too short (on word without context) or too long (text need to be cut into pieces). The solutions are resp. the use of a dictionary and existing python code.

Here is the proposed sequence:
1. tensorflow -> pytorch for [pipeline-diacritisation](https://github.com/Hamza5/Pipeline-diacritizer.git)
2. Benchmark results and make sure tensorflow =< pytorch... 
3. Fix special cases: sentences too long and too short
4. Integrate code into your solution
5. research phase: LSTM-> transformers, use BERT and attention ideas, among other ideas like alternative encoding, ...

timescale:
1. ~0.75 week
2. ~0.25 week
3. ~1 day
4. @ronaldtse  and @AhMohsen46 ...
5. ~2--4 weeks, we should be able to evaluate after ~1 week if chances of success...

**next meeting:** It seems that we are aligned on the first phase (tensorflow --> python), so let's meet the 3 of us after 1. and 2.? 

@AhMohsen46  is going to help me with expected problem in arabic...

**GPU**: I think I can do 1. and 2. without GPU's, will need GPU's from there.

**FALLBACK:** If everything above fails..., you can just use shakkala with tensorflow and available model, as I explained to @AhMohsen46 .

**Comment by ronaldtse** [comment]
Thanks guys! This looks like a good plan. After all this we might also be able to use AraBERT for named entity detection...

> next meeting: It seems that we are aligned on the first phase (tensorflow --> python), so let's meet the 3 of us after 1. and 2.?

Agree!

@AhMohsen46 can we also maintain a page that links to the related Arabic ML resources?

**Comment by gilgameshjw** [comment]
@AhMohsen46 I am happy to help for the last point.

For using AraBERT, obviously we won't be able to use it as it is as we are working without harakat, but there are several potential ideas to use it somehow to enrich the arabic without harakat.

**Comment by gilgameshjw** [comment]
Hello, I do think I have a running version in pytorch on my computer. After one epoch, results and score are promising.
I am finishing to wrap up a slightly better version of the scripts and a few plots and compute with the std benchmarks.

As mentionned above, it would be good to have a meeting with @ronaldtse  @AhMohsen46  at this point. 
There are various points to be addressed, among others:
* integration and production
* computing ressources (and costs?)
* potential improvements and schedule with that
* documentation and commit of code

I am available outside of tomorrow morning.

**Comment by ronaldtse** [comment]
The work will be continually developed at https://github.com/interscript/arabic-diacritization .

**Comment by gilgameshjw** [comment]
# Where are we?
## Benchmarks:
* benchmark [datasets](https://github.com/interscript/arabic-diacritization/tree/master/tests-benchmarks):
	* tashkeela benchmark data
	* ahmad practical examples
	* data arabic pointing
* Benchmarks are WER and DER with and without endings. 
* [helpers](https://github.com/interscript/arabic-diacritization/tree/master/tests-benchmarks) to clean up arabic from diacritics and produce scores


## Diacritization
While diacritization can be learned to high accuracy in general, any algorithm would struggle without context with "collisions" (words than can have several diacritizations). But one will need a hybrid approach:
1. Search
2. If no collision
	* NNets output
3. if collision
	* Business rules
4. Romanization

So the work will include both NNets and search dimensions.

## Diacritization :: Search & Romanization
To be designed/discussed with Ahmad!

## Diacritization :: NNets:
NNets are far better than business rules in term of performance and can be used without real understanding of the  language (sadly).
**Problem is that best models are too heavy for normal computers/CPU.**

We are interested in model learning the language at the character level since they are exepected to be better with unseen words (locations, names, ...).

To give an idea, the best NNets are currently around **DER** 1.79 **WER** 4.63 **DER-1** 1.15 **WER-1** 2.13, with DER-1/WER-1 for ignoring last character. With [paper](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9274427) and [code](https://github.com/almodhfer/Arabic_Diacritization) claiming the best results  (06.2021).



#### Players for GPU's:
* [deep diacritization](https://github.com/BKHMSI/deep-diacritization) 10.2020
* [Effective Deep Learning Models for Automatic Diacritization](https://github.com/almodhfer/Arabic_Diacritization) 03.2021

The later should have the best scores, but needs to be trained from scratch. 
Not all models seems to work straightaway but the CBHG architecture (in which we are very interested) does after a few tweaks.

#### Players for CPU's:
* One of the above with tweaks and speed up (less params)? The above are not working on normal CPU's as they are configured!
* [pipeline diacritizer](https://github.com/Hamza5/Pipeline-diacritizer) 
 the model is smaller than the above and faster, but better than shakkala by a margin. it has been ported to pytorch and benchmarked on small training sets.
 * [Shakkala](https://github.com/Barqawiz/Shakkala) This essentially is a fallback, tensorflow based library, the one to beat on GPU's. The full code is not available and as a result hard to replicate.

## Data
We use the data as in [paper](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9274427). Basically, [Tashkeela processed dataset](https://reader.elsevier.com/reader/sd/pii/S2352340917300112?token=0825CB9EEB2440D8305E7C29D43754FF3D8116640D81C44E53596B20E1903F3439F882498E5E8B8DC28E40CD67960471&originRegion=eu-west-1&originCreation=20210622143833), 
overall ~2'500'000 sentences.

# "Road to production"

#### 1. train and select best GPU
* compare with D2, D3 of deep-diacritization
*  train CBHG for few days on sagemaker on tashkeela
* comparison systematised on metrics and benchmarks specified above
* Make experiments decreasing sizes to increase speed, so that eventually we solve the point below
#### 2. train and select best CPU
* train pipeline GPU, pytorch
	* tweak code to add predictions on pipeline version
	* check results with full tashkeela
* compare with and hopefully crush Shakkala on benchmarks mentioned above

#### 3. Add text preprocessing and postprocessing
Most models need data clearning and preprocessing to remove non arabic symbols, decompose sentence into several ones if too long, etc...
One will then need to recompose it

#### 4. convert codes to ONNX
I had a chance to look at ONNX and pytorch examples, seems relatively straightforward...

#### 5. refactor & clean & document codes
... To have simpler codes to execute the diacritizations. Most of the codes above are research codes needing to be tweaked/adapted to prod.
As an example, it is non trivial to just diacritize text with most of them...

#### 6. add search and rules for single words
with Ahmad...

#### 7. Docker, kubernetes? environment? Integration to Ruby?
Here I will need Ahmad+others help.

#### 8. Feedbacks, more tests and improve
Ahmad and scores will help us to monitoring the points above.

# Deliveries:
Planning the following deliveries:
1. working, usable codes for diacritization with NNets:
	* training on large arabic datasets (with scripts+docu. to download them)
	* benchmarks: scientific and practical datasets with documented results on github
	* GPU (best model 2020--2021)
	* CPU (best smaller model > Shakkala)
	* ONNX
	* python 3.6
	* Documentation and good quality code for running NNets
	* Reports on experiments and training results
2. NNets + search as fallback for single words
	* as mentioned above, single words will need special handling
3. Integration, Docker or kubernetes container?

# Schedule
Here, below, this is an aggressive scenario to be discussed that I think is realistic and gives and evaluation of the work to be done.

#### 1. train and select best GPU
	I do need 2 instances of 4 gpus on it for this!
	Will be done in parallel of other tasks as some trainings 
	are > 1 days.
			~3.5 days:
			~1 days monitoring training
			~1 days tweaks on code 
			~1 days tests on smaller models
			~.5 days benchmarks and feedbacks
	23 -- 29 june
#### 2. train and select best CPU
	In parallel of the above:
			~2.5 days:
			~1 days to train pipeline
			~1 days to refactor and tweak pipeline
			~.5 days benchmark
	23 -- 29 june
#### 3. Add text preprocessing and postprocessing
			~1 days
			29 june
#### 4. convert codes to ONNX
			~1 days
			30 june
#### 5. refactor & clean & document codes
			~2 days
			1-2 july
#### 6. add search and rules for single words
			~2 days
			3-4 july
#### 7. Docker, kubernetes? environment? Integration to Ruby?
			~2 days for cpu and gpu
			5-6 july
#### 8. Feedbacks, more tests and improve
			~3 days
			7-10 july
#### Buffer, problems and some free days...
			~> 15 july

**Comment by ronaldtse** [comment]
@gilgameshjw the plan looks good, let's proceed!

Is it possible to prune the larger model down to the smaller model, e.g. https://stackoverflow.com/questions/62326683/prunning-model-doesnt-improve-inference-speed-or-reduce-model-size ?

Or is the better approach to train a smaller model from the beginning? I guess the time/cost is similar...

**Comment by gilgameshjw** [comment]
# To do list + approximate schedule

## NNets and Python Code
### update readme's, about usage and data
@gilgameshjw 
##### 5 July
### test nnets 
@gilgameshjw  @AhMohsen46 
##### 5-6 July
* test nnets on other datasets and check again with ahmad examples
* test nnets model on single words collisions/not collisions

### add business rules
@gilgameshjw  @AhMohsen46 
##### 7-9 july
* add business rules + search around existing code for single words

### add text pre/post-processing
##### 10-12
* add text length preprocessing // postprocessing (text too long must be decomposed, code exists in [test code](https://github.com/interscript/arabic-diacritization/tree/master/tests-benchmarks/helpers))

## Productionalisation
### python -> onnx+ruby test/play phase
@AhMohsen46 @gilgameshjw ?
##### ? July
* port onnx to ruby and test outputs (so that it matches python's)
* replicate secryst api, mocking pre- and post- processing
* take a more complex model, and activate pre-processing
	* i. calling python from ruby
	* ii. replacing function in ruby
### python -> onnx+ruby real phase
@AhMohsen46  @gilgameshjw  ?
##### ? July
Apply same on [real codebase](https://github.com/interscript/arabic-diacritization/tree/master/Arabic_Diacritization).
### python rules/code -> ruby
@AhMohsen46  @gilgameshjw ?
##### ? July
Replicate business rules and python preprocessing in ruby [real codebase](https://github.com/interscript/arabic-diacritization/tree/master/Arabic_Diacritization).

## Finalise
### Clean up, Readme's, ...
@gilgameshjw @AhMohsen46 
##### ? July

**Comment by gilgameshjw** [comment]
# TO DO for nnets prod in ruby
12.07.2021

- arabic -> ruby -> onnx -> diacritization (jair, 1 day) 
- add diacritizing files configs, case of long strings -> break them... (jair, 1 day)
- clean up & document (jair, 1 day)
- bug with some cpu's (jair ?)
- gem (ahmad, ?)

**Comment by gilgameshjw** [comment]
@ronaldtse @AhMohsen46 

**Errors types and behaviours should be specified and handled by interscript**. 

We need to agree on the following issues:

1. Is any input expected to produce an error, if yes, which kind of error?

2. If string entirely or partially non arabic, rebaba is expected to output it as it came, e.g. 
* example 1: "123 some arabic blabla @44 ##" --> "123 some diacritised arabic blabla @44 ##" 
* example 2: "" --> ""
* example 3: "$" --> "$"

Is that adequate?

**Comment by gilgameshjw** [comment]
using interscript as:
interscript ara_Arab2Latn_ALA_1997_diacritized_cpu.csv --system=bgnpcgn-ara-Arab-Latn-1956 --output=out_diacritized.csv

The mappings bgnpcgn-ara-Arab-Latn-1956 and alalc-ara-Arab-Latn-1997 are not outputting what is expected in the pairs.

**Comment by ronaldtse** [comment]
> 1. Is any input expected to produce an error, if yes, which kind of error?

This is a question for @AhMohsen46 

> 2. If string entirely or partially non arabic, rebaba is expected to output it as it came, e.g.
> 
> * example 1: "123 some arabic blabla @44 ##" --> "123 some diacritised arabic blabla @44 ##"
> * example 2: "" --> ""
> * example 3: "$" --> "$"

Agree.

> interscript ara_Arab2Latn_ALA_1997_diacritized_cpu.csv --system=bgnpcgn-ara-Arab-Latn-1956 --output=out_diacritized.csv

Can you provide some sample output for reference? Thanks.

**Comment by ronaldtse** [comment]
Closing this since it's done using Rababa. Thank you guys!
